src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php

```diff
@@ -57,36 +60,60 @@ public function validate($entity, Constraint $constraint)
             throw new ConstraintDefinitionException('At least one field has to be specified.');
         }
 
-        if (null === $entity) {
+        if (null === $object) {
             return;
         }
 
+        $entityClass = \get_class($object);
+
         if ($constraint->em) {
             $em = $this->registry->getManager($constraint->em);
 
             if (!$em) {
                 throw new ConstraintDefinitionException(sprintf('Object manager "%s" does not exist.', $constraint->em));
             }
         } else {
-            $em = $this->registry->getManagerForClass(\get_class($entity));
+            if (null !== $constraint->entityClass) {
+                $em = $this->registry->getManagerForClass($constraint->entityClass);
+            } else {
+                $em = $this->registry->getManagerForClass($entityClass);
+            }
 
             if (!$em) {
-                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($entity)));
+                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($object)));
             }
         }
 
-        $class = $em->getClassMetadata(\get_class($entity));
+        try {
+            $repository = $em->getRepository($entityClass);
```

Comment:
Shouldn't we use `$constraint->entityClass` here if possible?

Reply:
I need to check if the object being validated is an entity.
The `$constraint->entityClass` always have an entity class.
Maybe I could change `$entityClass` to `objectClass`? After all I changed `@param object $entity` to `@param object $object`

Reply:
```php
$entityClass = $constraint->entityClass ?? \get_class($object);
```

What I mean is couldn't we simply update line 67 to something like this and then use `$entityClass` in the rest of the method?

Reply:
@xabbuh Sadly, in lines 94-107 I need both values and I can't move it before line 69. The code use `$em` which is set after line 69.
I changed `$constraint->entityClass` to `$entityClass` after line 107.
I also simplified `this->registry->getManagerForClass` call.

Reply:
In this block, for simplicity, we should ONLY try to determine `$isEntity`. By also trying to set `$repository`, I think it confuses things. I would:

```php
try {
    $em->getRepository($value::class);
    $isEntity = true;
} catch (ORMMappingException|PersistenceMappingException) {
    $isEntity = false;
}
```

Or would this be enough?

```php
$isEntity = $em->contains($value::class)?
```

Also, calling thex variable `$isObjectEntity` or `$isValueEntity` might be even more clear :) 

Then, get the `$repository` always below this.

Reply:
I don't remember why I didn't use contains 3 years ago. During the weekend I will check it on real examples, not interfaces.

Reply:
Cool - thanks :). And more important for me is to ONLY set the `$isEntity` variable in this block. Then move figuring out the `$repository` until later... ideally moving all of the logic for getting the repository back the bottom - https://github.com/symfony/symfony/pull/38662/files#diff-fd1785451fbb1d6abe41ae644ddbe7b39b6f03f3ba31f7ce07a18ef376e61293L130 - as that would greatly reduce the diff. I'm a little obsessed with reducing the diff on this PR just because with the current set of changes, it's really hard to see & understand what's changed. It may turn out that the diff DOES need to be large like this, but I'm hoping we can shrink it down by avoiding unnecessary code movement. That'd make the PR much easier to merge 👍